miden-lib/asm/kernels/transaction/lib/memory.masm

```diff
@@ -607,6 +635,76 @@ end
 # ACCOUNT DATA
 # -------------------------------------------------------------------------------------------------
 
+#! Returns the pointer to the next vacant memory slot if the account was not loaded before, and the 
+#! pointer to the account data otherwise.
+#!
+#! Stack:  [foreign_account_id]
+#! Output: [was_loaded, ptr, foreign_account_id]
+#!
+#! Where: 
+#! - foreign_account_id is the ID of the foreign account whose procedure is going to be executed.
+#! - was_loaded is the binary flag indicating whether the foreign account was already loaded to the 
+#!   memory.
+#! - ptr is the memory pointer to the next empty memory slot or the memory pointer to the account 
+#!   data, depending on the value of the was_loaded flag.
+#!
+#! Panics if:
+#! - the ID of the provided foreign account equals zero.
+#! - the maximum allowed number of foreign account to be loaded (64) was exceeded.
+export.get_foreign_account_ptr
+    # check that foreign account id is not equal zero
+    dup neq.0 assert.err=ERR_FOREIGN_ACCOUNT_ID_IS_ZERO
+    # => [foreign_account_id]
+
+    # push the initial account data pointer
+    push.NATIVE_ACCOUNT_DATA_PTR
+    # => [curr_account_ptr, foreign_account_id]
+
+    # push the flag to enter the loop
+    push.1
+
+    while.true
+        # drop the flag left from the previous loop
+        movup.2 drop
+        # => [curr_account_ptr, foreign_account_id]
```

Comment:
Maybe I'm misunderstanding something, but wouldn't this drop whatever comes after the `foreign_account_id` on the stack in the very first loop iteration? And is that (not) a problem?

And this spawned a question for me. Suppose the above is true and a value is being dropped. This procedure is called from `start_foreign_context` which is executed in the root context via a syscall. The stack after the syscall is the same as before except that the items beyond the 16th are not accessible. Would it be fine for the procedure to drop some value that was on the stack before the syscall? I guess what I'm getting at is: Is there a calling convention that describes how procedures are supposed to manipulate or not manipulate the stack?

Reply:
>Maybe I'm misunderstanding something, but wouldn't this drop whatever comes after the foreign_account_id on the stack in the very first loop iteration? And is that (not) a problem?

That's a good catch, although _here_ it is not a problem.

First of all, I agree that this is not a clean implementation: ideally I should push some default flag (e.g. 0) to make sure I don't drop a potentially important value from the stack. I don't do that because here the `syscall`'ed `start_foreign_context` procedure guarantee that at the moment of `get_foreign_account_ptr` being executed the stack looks like so:
```
[foreign_account_id, pad(15)]   # where `pad` is in fact just a `0` 
```
This is achieved because I pad the stack before I `syscall`'ed the `start_foreign_context` ([here](https://github.com/0xPolygonMiden/miden-base/blob/b8d9541d17d40561a1f1c623e2d771cde799553c/miden-lib/asm/miden/tx.masm#L177)).

`get_foreign_account_ptr` is a specific procedure which should be used only in the `start_foreign_context`, so that's why I can slightly optimize the execution here. 

>Would it be fine for the procedure to drop some value that was on the stack before the syscall? I guess what I'm getting at is: Is there a calling convention that describes how procedures are supposed to manipulate or not manipulate the stack?

Great question! it is indeed a problem if a `call`'ed procedure drops something which it wasn't supposed to drop. To prevent such situations Bobbin suggested to pad the stack before `call`, `syscall` and `dyncall` invocations. This approach is described in detail in the related issue: https://github.com/0xPolygonMiden/miden-base/issues/685

Reply:
Awesome, thanks for the extensive answer, very clear. It's really good that there's a convention for this!